File: src/main/java/no/fintlabs/model/configuration/dtos/FromCollectionMappingDto.java
package no.fintlabs.model.configuration.dtos;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.extern.jackson.Jacksonized;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Getter
@Builder
@EqualsAndHashCode
@Jacksonized
public class FromCollectionMappingDto<T> {

    public FromCollectionMappingDto(List<String> instanceCollectionReferencesOrdered, T elementMapping) {
        this.instanceCollectionReferencesOrdered = Optional.ofNullable(instanceCollectionReferencesOrdered).orElse(new ArrayList<>());
        this.elementMapping = elementMapping;
    }

    @NotEmpty
    private final List<@NotBlank String> instanceCollectionReferencesOrdered;

    @Valid
    @NotNull
    private final T elementMapping;

}

File: src/main/kotlin/no/novari/flyt/configuration/model/configuration/dtos/FromCollectionMappingDto.kt
package no.novari.flyt.configuration.model.configuration.dtos

import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotEmpty
import jakarta.validation.constraints.NotNull

data class FromCollectionMappingDto<T>(
    @field:NotEmpty
    var instanceCollectionReferencesOrdered: MutableList<@NotBlank String> = mutableListOf(),
    @field:Valid
    @field:NotNull
    var elementMapping: T? = null,
) {
    companion object {
        @JvmStatic
        fun <T> builder(): Builder<T> = Builder()
    }

    class Builder<T> {
        private var instanceCollectionReferencesOrdered: MutableList<String> = mutableListOf()
        private var elementMapping: T? = null

        fun instanceCollectionReferencesOrdered(instanceCollectionReferencesOrdered: List<String>?) =
            apply {
                this.instanceCollectionReferencesOrdered =
                    instanceCollectionReferencesOrdered?.toMutableList() ?: mutableListOf()
            }

        fun elementMapping(elementMapping: T?) = apply { this.elementMapping = elementMapping }

        fun build(): FromCollectionMappingDto<T> =
            FromCollectionMappingDto(
                instanceCollectionReferencesOrdered = instanceCollectionReferencesOrdered,
                elementMapping = elementMapping,
            )
    }
}
